cmd/starter/main.go

```diff
@@ -130,6 +131,26 @@ func fetchBuildJob() {
 	}
 	repo := job.Repository
 
+	// Aquire lock to ensure one repo is not built twice
+	lockID := uuid.NewV4().String()
+	redisClient.Set(fmt.Sprintf("project::%s::build-lock", repo), lockID, 1800, 0, false, true)
+
+	lock, err := redisClient.Get(fmt.Sprintf("project::%s::build-lock", repo))
+	if err != nil {
+		log.WithFields(logrus.Fields{
+			"error": err,
+		}).Error("Unable to fetch lock state")
+		return
+	}
+
+	if string(lock) != lockID {
+		// If build-job is currently locked but not by us, try again later
+		queueEntry, err := job.ToByte()
+		orFail(err)
+		redisClient.RPush("build-queue", string(queueEntry))
+		return
+	}
```

Comment:
Ohh maybe, just caught the `mustNotExists=true` in the `Set` call. http://redis.io/commands/set  claims something sort of atomic. maybe it returns an error if it fails?